docs2/site/docs/guides/complexity-analyzer.md

```diff
@@ -0,0 +1,346 @@
+and provide your own validation of the calculated depth:
+
+```csharp
+complexityConfig.MaxDepth = null;
+complexityConfig.ValidateComplexityDelegate = async (context, complexity, depth) =>
+{
+    // for non-introspection queries, validate the depth
+    if (complexity > 0 && depth > 10)
+    {
+        context.ReportError(new ExecutionError("Query depth is too high."));
+    }
+};
+```
+
+### 3. Estimating Computing Power (Database Processing Time)
+
+Another use case for the complexity analyzer is to estimate the computing power required to process a query.
+You can configure the impact for object fields to estimate the database processing time by setting a custom default
+object impact or configuring the impact for specific fields. The below examples assume that the scalar impact is 1,
+but you may wish to adjust this to zero if scalar fields do not require consequential processing time.
+
+#### Configuring Impact for Object Fields
+```csharp
+// Set higher impact for field resolvers that require more processing time
+usersField.WithComplexityImpact(50);
+```
+
+#### Setting a Custom Default Object Impact
+```csharp
+// Set default for object fields (assumed to need to load from a database)
+complexityConfig.DefaultObjectImpact = 20;
+```
+
+#### Sample GraphQL Request and Computed Complexity
+
+```graphql
+query {            #  impact   multiplier   total impact   child multiplier   depth
+  users {          #    50          1            50                  5          1
+    id             #     1          5            55                             2
+    posts {        #    20          5           155                  5          2
+      id           #     1         25           180                             3
+      comments {   #    20         25           680                  5          3
+        id         #     1        125           805                             4
+      }
+    }
+  }
+}
+```
+
+The above query will have the following complexity calculation:
+- Maximum Depth: 4 (users -> posts -> comments -> id)
+- Total Complexity: 805
+
+### 4. Logging Complexity Results
+
+In addition to validation, the `ValidateComplexityDelegate` property allows you to log complexity results
+for monitoring or analysis. Please note that this delegate runs after the maximum depth and complexity checks,
+so it cannot be used to log queries that fail these checks.
+
+```csharp
+complexityConfig.ValidateComplexityDelegate = async (context, complexity, depth) =>
+{
+    // RequestServices may be used to access scoped services within the DI container
+    var logger = context.RequestServices!.GetRequiredService<ILogger<MySchema>>();
+    logger.LogInformation($"Query Complexity: {complexity}, Depth: {depth}");
+};
+```
+
+To log failed complexity checks, set the `MaxDepth` and `MaxComplexity` properties to `null` and provide
+your own validation delegate:
```

Comment:
Right now we have:

```cs
public Func<ValidationContext, double, int, Task>? ValidateComplexityDelegate { get; set; }
```

Rather than add yet another delegate, how about if this becomes:

```cs
public Func<ComplexityResult, Task>? ValidateComplexityDelegate { get; set; }

public class ComplexityResult
{
    public ValidationContext ValidationContext { get; }
    public ComplexityOptions ComplexityOptions { get; }
    public double TotalImpact { get; }
    public int MaxDepth { get; }
    public ComplexityError? Error { get; set; } // null if no error
}

// when validation occurs, initialize the ComplexityResult class, setting the Error
//   property if the impact or depth has been exceeded; then call the delegate.
// if the error has been cleared, continue as if no problem, otherwise report
//   the error within the Error property
```

Reply:
The `Error` property could be named something else, of course, like `Result`.  But as the class is `ComplexityResult`, I probably wouldn't.  Also we could allow `ValidationError` instead of `ComplexityError`.

Reply:
How about getting inspiration form [CookieAuthenticationEvents](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.authentication.cookies.cookieauthenticationevents?view=aspnetcore-8.0)? Every method in this class receives its own context type. The `ValidatePrincipal` receives [CookieValidatePrincipalContext](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.authentication.cookies.cookievalidateprincipalcontext?view=aspnetcore-8.0). There is a property called `Principal` and two methods: `RejectPrincipal` and `ReplacePrincipal`. When you see methods like Reject and Replace, you intuitively understand how to use the context and don't have to guess what to do with the `Principal` property (like setting it to `null`, for example).

In our case, it can be `ValidateComplexityContext`, with `ValidationResult` property and `FailValidation(ValidationError)` + `RejectValidation`
```csharp
public class ValidateComplexityContext
{
    public ValidationContext ValidationContext { get; }
    public ComplexityOptions ComplexityOptions { get; }
    public double TotalImpact { get; }
    public int MaxDepth { get; }
    public ValidationError? ValidationResult { get; private set;}

    public void FailValidation(ValidationError error) => ValidationResult = error;
    public void RejectValidation() => ValidationResult = null; // better name?
}
```

Reply:
Btw, can it return `ValueTask` instead of `Task`?

Reply:
I think it would be better to mesh with code already familiar to GraphQL.NET users rather than copy a design from another class.  I for one have no familiarity with `CookieAuthenticationEvents` or `CookieValidatePrincipalContext`, and none of my devs do either.  I also don't find `Fail` and `Reject` to be clear; they seem to be synonyms.

We could also forgo the `ValidationResult` field altogether and do this:

```cs
public Func<ComplexityResult, Task> ValidateComplexityDelegate { get; set; } = result =>
{
    // default implementation
}


public class ComplexityResult
{
    public ValidationContext ValidationContext { get; }
    public ComplexityOptions ComplexityOptions { get; }
    public double TotalImpact { get; }
    public int MaxDepth { get; }
}
```

The default implementation calls `ValidationContext.AddError` to add an error if the impact or depth is exceeded.  However, this is a pain when overriding as now the default logic needs to be re-implemented, and/or the code needs to examine `ValidationContext` to determine if any complexity errors were added.  So I really don't like this.

Yet another option is this:

```cs
public Func<ComplexityResult, ValueTask<ComplexityError?>> ValidateComplexityDelegate { get; set; } = result =>
{
    // default implementation
}

public struct ComplexityResult
{
    public ValidationContext ValidationContext { get; }
    public ComplexityOptions ComplexityOptions { get; }
    public double TotalImpact { get; }
    public int MaxDepth { get; }
}
```

User-defined delegates can call the original implementation and then modify the result.  One benefit is that no allocations are typically necessary.  Again, this is more difficult to utilize than the approach used by `UnhandledErrorDelegate`; the simplest implementation would be:

```cs
var originalDelegate = options.ValidateComplexityDelegate;
options.ValidateComplexityDelegate = async (result) =>
{
    var error = await originalDelegate(result);
    if (error != null)
        // log error
    return error;
};
```

Whereas with the suggestion above, it would be:

```cs
options.ValidateComplexityDelegate = async (result) =>
{
    if (result.Error != null)
        // log error
}
```

I think the majority of the use case here is either:
1. Logging analysis results (perhaps only failures, perhaps all for statistical analysis)
2. Applying additional throttling restrictions based on the calculated total impact

As such, we should focus on those use cases, making it easy for the developer.

Reply:
> Btw, can it return `ValueTask` instead of `Task`?

There is little-to-no value when not returning `Task<T>`, as the returned `Task` will be a cached instance anyway.  I've tried to mask all use of `ValueTask` from the developer whenever possible.  However, here it is critical to use `ValueTask` for `CalculateComplexityAsync` as this is executed for each field and must return a value in each case.  Some of the GraphQL.NET interfaces do use `ValueTask` when `Task` would have been equally sufficient, just to maintain parity with the rest of the methods within the same class.

Reply:
See: "Should I use ValueTask for new asynchronous APIs?" here https://devblogs.microsoft.com/dotnet/understanding-the-whys-whats-and-whens-of-valuetask/

Reply:
> I think it would be better to mesh with code already familiar to GraphQL.NET

Completely understandable. Then this one is the best of all the proposals because it's similar to `UnhandledExceptionDelegate`
```c#
public Func<ComplexityResult, Task>? ValidateComplexityDelegate { get; set; }
```
Just the names. Again, for consistency with `UnhandledExceptionDelegate`
```c#
public Func<ComplexityValidationContext, Task>? ValidateComplexityDelegate { get; set; }

public class ComplexityValidationContext
{
    public ValidationContext ValidationContext { get; }
    public ComplexityOptions ComplexityOptions { get; }
    public double TotalImpact { get; }
    public int MaxDepth { get; }
    public ValidationError? Error { get; set; } // null if no error
}
```

Reply:
> See: "Should I use ValueTask for new asynchronous APIs?" here https://devblogs.microsoft.com/dotnet/understanding-the-whys-whats-and-whens-of-valuetask/

But then there is another opinion from another shinning name in the .NET world
[Prefer ValueTask to Task, always; and don't await twice](https://blog.marcgravell.com/2019/08/prefer-valuetask-to-task-always-and.html)

Reply:
Done, but I need to update the docs.

Reply:
Docs updated